src/lem/interpreter.rs

```diff
@@ -1,206 +1,253 @@
+                preimages.hash4.push(preimg_ptrs);
+                Ok(())
+            }
+            LEMOP::Unhash2(preimg, img) => {
+                let img_ptr = img.get_ptr(bindings)?;
+                let Some(idx) = img_ptr.get_index2() else {
+                    bail!("{img} isn't a Tree2 pointer");
+                };
+                let Some((a, b)) = store.fetch_2_ptrs(idx) else {
+                    bail!("Couldn't fetch {img}'s children")
+                };
+                let vec = [*a, *b];
+                for (mptr, ptr) in preimg.iter().zip(vec.iter()) {
+                    bind(bindings, mptr.clone(), *ptr)?;
                 }
-                LEMOP::Unhash3(tgts, src) => {
-                    let src_ptr = src.get_ptr(&ptrs)?;
-                    let Some(idx) = src_ptr.get_index3() else {
-                        bail!(
-                            "{} isn't a Tree3 pointer",
-                            src.name()
-                        );
-                    };
-                    let Some((a, b, c)) = store.fetch_3_ptrs(idx) else {
-                        bail!("Couldn't fetch {}'s children", src.name())
-                    };
-                    insert_into_ptrs(&mut ptrs, tgts[0].name().clone(), *a)?;
-                    insert_into_ptrs(&mut ptrs, tgts[1].name().clone(), *b)?;
-                    insert_into_ptrs(&mut ptrs, tgts[2].name().clone(), *c)?;
+                preimages.hash2.push(vec.to_vec());
+                Ok(())
+            }
+            LEMOP::Unhash3(preimg, img) => {
+                let img_ptr = img.get_ptr(bindings)?;
+                let Some(idx) = img_ptr.get_index3() else {
+                    bail!("{img} isn't a Tree3 pointer");
+                };
+                let Some((a, b, c)) = store.fetch_3_ptrs(idx) else {
+                    bail!("Couldn't fetch {img}'s children")
+                };
+                let vec = [*a, *b, *c];
+                for (mptr, ptr) in preimg.iter().zip(vec.iter()) {
+                    bind(bindings, mptr.clone(), *ptr)?;
                 }
-                LEMOP::Unhash4(tgts, src) => {
-                    let src_ptr = src.get_ptr(&ptrs)?;
-                    let Some(idx) = src_ptr.get_index4() else {
-                        bail!(
-                            "{} isn't a Tree4 pointer",
-                            src.name()
-                        );
-                    };
-                    let Some((a, b, c, d)) = store.fetch_4_ptrs(idx) else {
-                        bail!("Couldn't fetch {}'s children", src.name())
-                    };
-                    insert_into_ptrs(&mut ptrs, tgts[0].name().clone(), *a)?;
-                    insert_into_ptrs(&mut ptrs, tgts[1].name().clone(), *b)?;
-                    insert_into_ptrs(&mut ptrs, tgts[2].name().clone(), *c)?;
-                    insert_into_ptrs(&mut ptrs, tgts[3].name().clone(), *d)?;
+                preimages.hash3.push(vec.to_vec());
```

Comment:
This is *extremely* confusing. Let me describe what I *think* I understand in this code, here and below.

Here, we have a field killed `hash3` of `preimages`. This contains preimages that are composed of 6 field elements.

Reply:
Would changing it to `preimages.hash3_ptrs` help? We don't want to afford calls to poseidon hashing during interpretation. What we're collecting here are pointers, not the field elements

Reply:
Yes, that would avoid the confusing name collision.